### PR TITLE
feat: Add support for event-driven catalog sync

### LINKED ISSUE #3290: [Event Driven Catalog Sync] Implement event driven catalog provider  on Backstage
### Description

Make the OpenChoreo catalog provider event-driven by consuming the webhook notifications dispatched by the new OpenChoreo watcher  using event primitives provided by Backstage such as `@backstage/plugin-events-backend`

### Version

v1.1.0

### Related Issue

https://github.com/openchoreo/openchoreo/issues/3155

### PR DESCRIPTION
**Pull request by stefinie123**
## Purpose

Today the Backstage catalog only learns about OpenChoreo state via a periodic full sync (default 30s), so the developer portal can be stale by up to one full sync interval after any change. The sync also re-fetches every Project, Component, Workload, etc. on every tick, which scales poorly as orgs grow.                                                                                                                                                                                    
                                                                                                                                                                                                                                                     
This PR makes the catalog event-driven: changes in OpenChoreo propagate to Backstage within seconds via webhooks pushed by the new event-forwarder, while the periodic sync is retained at a reduced cadence (300s) as a consistency safety net

Resolves https://github.com/openchoreo/openchoreo/issues/3290

## Goals

Support event driven catalog sync in Backstage

## Approach

Wire the new OpenChoreo event-forwarder into Backstage using @backstage/plugin-events-backend, and add a webhooks ingress + dispatch layer that turns those events into per-resource catalog mutations. The events backend exposes the HTTP topic (/api/events/http/openchoreo) the forwarder POSTs to, and routes every received event to in-process subscribers via the events service. The existing periodic full sync is preserved unchanged; the event path simply applies deltas in between    sync ticks

### New components
                                                                                                                                                                                                                                                     
  - `OpenChoreoEventRouter` — a SubTopicEventRouter subscribed to the openchoreo topic (fed by the events HTTP plugin) that republishes each event to openchoreo.<kind> sub-topics, so the entity provider can subscribe selectively per resource kind.
  - `EventDeltaApplier` — the heart of the event path. For each event the entity provider receives, it fetches the affected resource(s) from the OpenChoreo REST API and applies a delta mutation. Reuses the same translators as the full sync, so
  both paths converge to identical entities.  
  - `SystemEntityProcessor` — emits the Project↔DeploymentPipeline relation pair from the Project (System) side using spec.deploymentPipelineRef, replacing the previous inverted-index pattern that left stale relations on pipeline reassignment.    
                                              
  Notable mechanics                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                   
  - Workload events resolve the owning Component dynamically. Workload metadata.name and component name don't have to match — the dispatcher fetches the workload, reads spec.owner.componentName, and refreshes that Component (re-using the        
  already-fetched workload to avoid a second lookup).                                                                                                                                                                                              
  - Workload deletion uses an annotation-based catalog query. Component and API entities carry openchoreo.io/workload=<workloadName> (set both in the event path and the full sync). On workload-delete the dispatcher queries the catalog by that   
  annotation, removes the matched API entities, and refreshes the parent Component so its providesApis/consumesApis clear.                                                                                                                           
  - Periodic sync frequency moved 30s → 300s. With the event path landing within seconds, the full sync's role shifts from primary-driver to safety-net.
  - Feature gate. Setting openchoreo.events.enabled: false skips the event router and sub-topic subscriptions cleanly, leaving the provider in poll-only mode (mirrors the OpenChoreo Helm eventForwarder.enabled flag).                             
                                                                                                                                                                                                                                                     
  Removed                                 
                                                                                                                                                                                                                                                     
  - ComponentWorkflowEntityV1alpha1 kind + processor — the embedded workflow data is now an annotation on the Component entity, no separate kind needed.                                                                                             
  - DeploymentPipeline.spec.projectRefs inverted index — the relation lives on the Project side via SystemEntityProcessor.                                                                                                                           
                                                                                                                                                                                                                                                    

## User stories

> Summary of user stories addressed by this change>

## Release note

> Brief description of the new feature or bug fix as it will appear in the release notes

## Documentation

> Link(s) to product documentation that addresses the changes of this PR. If no doc impact, enter “N/A” plus brief explanation of why there’s no doc impact

## Training

> Link to the PR for changes to the training content in https://github.com/wso2/WSO2-Training, if applicable

## Certification

> Type “Sent” when you have provided new/updated certification questions, plus four answers for each question (correct answer highlighted in bold), based on this change. Certification questions/answers should be sent to certification@wso2.com and NOT pasted in this PR. If there is no impact on certification exams, type “N/A” and explain why.

## Marketing

> Link to drafts of marketing content that will describe and promote this feature, including product page changes, technical articles, blog posts, videos, etc., if applicable

## Automation tests

- Unit tests
- 
109 unit tests across EventDeltaApplier (29 — dispatch table, workload-event handling, catalog-annotation deletion path, race + cold-start edge cases), OpenChoreoEventRouter (4), processors (55, including new SystemEntityProcessor coverage),  and OpenChoreoEntityProvider (21).

- Integration tests
  > Details about the test cases and coverage

## Security checks

- Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
- Ran FindSecurityBugs plugin and verified report? yes/no
- Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

## Samples

> Provide high-level details about the samples related to this feature

## Related PRs

> List any other related PRs

## Migrations (if applicable)

> Describe migration steps and platforms on which migration has been tested

## Test environment

> List all JDK versions, operating systems, databases, and browser/versions on which this feature/fix was tested

## Learning

> Describe the research phase and any blog posts, patterns, libraries, or add-ons you used to solve the problem.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added event-driven catalog synchronization for real-time entity updates via webhooks.

* **Changes**
  * Default polling frequency increased from 30 to 300 seconds.
  * System entities now manage deployment pipeline relationships (previously via ComponentWorkflow).
  * Removed ComponentWorkflow entity type.
  * Updated DeploymentPipeline entity model for improved relationship tracking.

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/openchoreo/backstage-plugins/pull/514)

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: plugins/catalog-backend-module-openchoreo/src/provider/EventDeltaApplier.ts [hunk 1]
```diff
@@ -0,0 +1,1235 @@
...
+            endpoints: schemaEndpoints,
+            namespaceName: ns,
+            projectName,
+            owner: resolvedOwner,
+            locationKey: this.locationKey(),
+            workloadName,
+          })
+        : [];
+
+    await this.upsertEntities([componentEntity, ...apiEntities]);
```
[the remark this answers is not in the document]
Comment (stefinie123) [reply plugins/catalog-backend-module-openchoreo/src/provider/EventDeltaApplier.ts:628]: Deferring this to the periodic full sync, deliberately. Endpoint removal/rename is uncommon in practice, and the user-visible impact (a brief "ghost" API entity in the catalog for up to one full-sync interval, currently 300s) is small and self-healing. Adding a catalog query on every refreshComponent call to detect the rare stale-entity case would pay constant per-event I/O for a rare condition, and it'd break the symmetry we've established elsewhere in the event path: every other deferred edge case (cold-start race in handleWorkloadDeleted, concurrent create-vs-delete races, etc.) already relies on the next full sync's type: 'full' mutation as the safety net.                                                     
                                                                                                                                                                                                                                                     
  If we see real evidence in production that the ghost-API window matters to users, the annotation-query fix you suggested is the right shape — we'd add it then.

File: packages/backend/src/index.ts [hunk 2]
```diff
@@ -28,7 +28,6 @@ backend.add(
       app.use(middleware.helmet());
       app.use(middleware.cors());
       app.use(middleware.compression());
-      app.use(middleware.logging());
```
Comment (LakshanSS) [inline packages/backend/src/index.ts:31]: logging middleware is removed. Is this intentional?

File: app-config.yaml [hunk 3]
```diff
@@ -166,7 +166,7 @@ openchoreo:
   defaultOwner: 'openchoreo-users'
 
   schedule:
-    frequency: 30 # seconds between runs (default: 30)
+    frequency: 300 # seconds between runs (default: 300, the periodic full sync acts as a safety net behind the event-forwarder's real-time deltas)
```
Comment (LakshanSS) [inline app-config.yaml:169]: Will there be any scenario where an event could be missed and only picked up by the 300s full sync? If so, is 300s an acceptable delay for those cases?

File: packages/backend/src/index.ts (hunk 2 above)
Comment (stefinie123) [reply packages/backend/src/index.ts:31]: Yes this was intentional. In line 37 the `applyDefaults()` function anyways applies the logging middleware. 
With the previous code many of the logs were duplicating, and while debugging the issue was able to identify this duplicate apply. Removing this fixed the duplicate logs issue.

File: app-config.yaml (hunk 3 above)
Comment (stefinie123) [reply app-config.yaml:169]: Any network issues or issues in the event forwarder side could result in missed events, and those will only be captured by the full sync. However, with this PR and previous PR to OC we have made the full sync time configurable via helm values. Hence if someone feels like the delay is too much it is always configurable.

Additional we will be also introducing a retry mechanism to event-forwarder side, to reduce the probability of lost events.

File: packages/backend/src/index.ts (hunk 2 above)
Comment (LakshanSS) [reply packages/backend/src/index.ts:31]: Ack

File: app-config.yaml (hunk 3 above)
Comment (LakshanSS) [reply app-config.yaml:169]: Ack

**Review by LakshanSS — APPROVED** [review]
LGTM

**Review by kaviththiranga — APPROVED** [review]
@stefinie123 we should update docs as well
